c10/core/TensorImpl.cpp

```diff
@@ -537,17 +538,24 @@ template <typename VariableVersion>
 c10::intrusive_ptr<TensorImpl> TensorImpl::shallow_copy_and_detach_core(
     VariableVersion&& version_counter,
     bool allow_tensor_metadata_change) const {
-  if (key_set_.has(DispatchKey::Python) &&
+  c10::intrusive_ptr<TensorImpl> r;
+  const auto& maybe_torch_dispatch_mode_state =
+      c10::impl::TorchDispatchModeTLS::get_state();
+  // TODO: do we have to exclude after Python dispatch key set?
+  if (maybe_torch_dispatch_mode_state) {
+    r = maybe_torch_dispatch_mode_state->pyinterpreter()->detach(this);
+  } else if (
+      key_set_.has(DispatchKey::Python) &&
       !c10::impl::tls_is_dispatch_key_excluded(DispatchKey::Python)) {
-    auto r = pyobj_interpreter_.load(std::memory_order_acquire)->detach(this);
-    if (r) {
-      r->set_version_counter(std::forward<VariableVersion>(version_counter));
-      r->set_allow_tensor_metadata_change(allow_tensor_metadata_change);
-      return r;
-    }
-    // otherwise just copy the TensorImpl and not the PyObject.  Since
-    // the interpreter is dead no one can call us out on it
+    r = pyobj_interpreter_.load(std::memory_order_acquire)->detach(this);
+  }
+  if (r) {
+    r->set_version_counter(std::forward<VariableVersion>(version_counter));
+    r->set_allow_tensor_metadata_change(allow_tensor_metadata_change);
+    return r;
```

Comment:
I think we're going to need to make a similar change for the other tensor subclasses that override shallow_copy_and_detach().

`FunctionalTensorWrapper` ([here](https://github.com/pytorch/pytorch/blob/d4bd88b64be3d7156cf8617e33faae8eca307ce4/aten/src/ATen/FunctionalTensorWrapper.cpp#L300))

(more recently) `NestedTensorImpl` ([here](https://github.com/pytorch/pytorch/blob/d4bd88b64be3d7156cf8617e33faae8eca307ce4/aten/src/ATen/NestedTensorImpl.cpp#L249))

Although coming up with a better pattern to avoid the code dup'ing would probably be better longterm.

Reply:
I feel like we've been tiptoe-ing around the actual issue, which is that shallow_copy_and_detach is not a dispatch-able operation. If it were, then everything would just work out. This has led to issues in the past, like https://github.com/pytorch/pytorch/issues/71725 and https://github.com/pytorch/pytorch/issues/74991 (which is exactly what you're referring to, @bdhirsh).

Reply:
BTW, having the tensor subclasses properly wrap this will only be necessary when we want to directly trace operations involving those tensors. NestedTensorImpl will need it, but FunctionalTensorWrapper will not (we always trace "below" functional tensor wrapper).

Reply:
> NestedTensorImpl will need it, but FunctionalTensorWrapper will not (we always trace "below" functional tensor wrapper)

I think FunctionalTensorWrapper will for LTC/XLA (although to be fair, that's assuming that the way XLA integrates with AOTAutograd-style tracing is using lazy tensor infra)

Reply:
> This has led to issues in the past, like https://github.com/pytorch/pytorch/issues/71725 and https://github.com/pytorch/pytorch/issues/74991 (which is exactly what you're referring to, @bdhirsh). This probably won't happen soon, but do you think it would resolve the long-term problem?

Those gh issues were what I had in my head when making that comment :) Agreed they would fix this issue.

Reply:
It doesn't feel like, to me, we will be able to get rid of `shallow_copy_and_detach` as a concept on TensorImpl though; the proposed fix would simply to make sure you only ever got here via a dispatching operation.

Reply:
I was thinking more that that would give us a place to shunt some of this boilerplate code so it can be re-used (so at least, we can only have one copy of the code that knows to call torch_dispatch for all the different subclasses).

Reply:
Yes. The point, though, is that `shallow_copy_and_detach` would remain, but it would not be dispatched. (And the common code goes in detach)